### PR TITLE
Add support for Processing Instruction element

### LINKED ISSUE #119: Port #111 Add support for XML processing instructions in <content> to #102

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
related to: https://github.com/openstax-poland/adaptarr-front/issues/512 

Fixes #119

### CODE REVIEW

File: src/interfaces/processingInstruction.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Element } from 'slate'
+
+/** Processing Instruction */
+export interface ProcessingInstruction extends Element {
+    type: 'pi'
```
Comment (aiwenar) [inline src/interfaces/processingInstruction.ts:5]: I would prefer a less ambiguous name, such as `processing_instruction`.

File: src/cnxml/jsx.ts [hunk 2]
```diff
@@ -185,6 +186,11 @@ function finishElement(renderer: Renderer, element: Element, out: globalThis.Ele
             out.append(renderer.doc.importNode(child, true))
         } else if (typeof child === 'string') {
             out.append(child)
+        } else if (child.name.local === 'pi') {
+            out.append(document.createProcessingInstruction(
+                child.attributes.target as string,
+                child.attributes.value as string,
+            ))
```
Comment (aiwenar) [inline src/cnxml/jsx.ts:193]: This would prevent us from ever using an element named `pi`. It also confuses processing instructions with elements. Support for processing instructions should be more explicit:

1. Add new type for processing instructions:

   ```ts
   /** A processing instruction */
   export interface ProcessingInstruction {
      /** PI's target */
      target: string
      /** PI's value */
      value: string
   }
   ```

2. Add it as a possible `Node`:

   ```diff
    /** Any value that can be used as child of a JSX element */
   -export type Node = Element | globalThis.Node | string | Node[] | null
   +export type Node = Element | ProcessingInstruction | globalThis.Node | string | Node[] | null
   ```

3. Handle this new type in `renderChild`

   ```diff
    if (child instanceof globalThis.Node) {
       out.append(renderer.doc.importNode(child, true))
    } else if (typeof child === 'string') {
       out.append(child)
   +} else if (typeof child.target === 'string') {
   +   out.append(document.createProcessingInstruction(child.target, child.value))
    } else {
       out.append(renderElement(r, child))
    }
   ```

File: src/cnxml/se.tsx [hunk 3]
```diff
@@ -496,6 +497,19 @@ function mediaItem(
...
+    node: ProcessingInstruction,
+    attrs: CommonAttrs,
+    children: JSX.Node,
+): JSX.Node {
+    return <pi
+        target={node.target}
+        value={node.value}
+        {...attrs}>
+        {children}
+    </pi>
```
Comment (aiwenar) [inline src/cnxml/se.tsx:510]: Following my above comment about not confusing processing instructions with elements, this should be

```
return { target: node.target, value: node.value }
```

Regardless, adding `{...attrs}` and `{children}` here is confusing, as they will be ignored.

File: test/util/cnxml.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ export default global.cnxml = function cnxml(fragments) {
         throw new Error('cnxml`` does not support interpolation')
     }
 
-    return `<document xmlns="http://cnx.rice.edu/cnxml" cnxml-version="0.7" id="test" module-id="test">
+    return `<document xmlns="http://cnx.rice.edu/cnxml" cnxml-version="0.7" id="test" module-id="test" lang="en">
```
Comment (aiwenar) [inline test/util/cnxml.js:6]: Why is this change made here?

**Review by aiwenar — CHANGES_REQUESTED** [review]

File: test/util/cnxml.js (hunk 4 above)
Comment (PiotrKozlowski) [reply test/util/cnxml.js]: I've forgotten to remove this. I was trying to figure out why tests are failing

File: src/cnxml/jsx.ts [hunk 5]
```diff
@@ -30,8 +30,28 @@ export interface Element {
...
+    node && typeof (node as ProcessingInstruction).target === 'string')
+
+export const nodeIsNotJSXElement = (
+    node: Node,
+// eslint-disable-next-line max-len
+): node is ProcessingInstruction | globalThis.Node | string | Node[] | null => typeof node === 'string'
+        || node === null
+        || Array.isArray(node)
+        || node instanceof globalThis.Node
+        || nodeIsProcessingInstruction(node)
```
Comment (aiwenar) [inline src/cnxml/jsx.ts:54]: Please use normal functions, and preferably a namespace:

```typescript
const Node = {
    isProcessingInstruction(...)

   ...
}
```

File: src/cnxml/se.tsx [hunk 6]
```diff
@@ -401,14 +401,14 @@ function docref(node: DocumentReference, attrs: CommonAttrs, children: JSX.Node)
...
             return child.map(mapChild)
         }
 
+        if (JSX.nodeIsNotJSXElement(child)) {
+            return child
+        }
+
         if (child.name.namespace === JSX.CNXML_NAMESPACE
         && child.name.local === 'figure') {
             return {
```
Comment (aiwenar) [inline src/cnxml/se.tsx:414]: This change is unrelated to the PR.

File: src/cnxml/se.tsx [hunk 7]
```diff
@@ -440,12 +440,12 @@ function link(node: Link, attrs: CommonAttrs, children: JSX.Node): JSX.Node {
...
+
+        if (JSX.nodeIsNotJSXElement(child)) {
             return child
         }
 
-        if (Array.isArray(child)) return child.map(mapChild)
-
         if ((child.name.namespace == null || child.name.namespace === JSX.CNXML_NAMESPACE)
         && child.name.local === 'list') {
             return <item xmlns={JSX.CNXML_NAMESPACE}>{child}</item>
```
Comment (aiwenar) [inline src/cnxml/se.tsx:451]: This change is unrelated to the PR.

**Review by aiwenar — CHANGES_REQUESTED** [review]

File: src/cnxml/se.tsx (hunk 6 above)
Comment (PiotrKozlowski) [reply src/cnxml/se.tsx]: previous `if` statement was not including `ProcessingInstruction` so it was throwing an error since `child.name....` wasn't accessible on `child: Element | ProcessingInstruction`

File: src/cnxml/se.tsx (hunk 7 above)
Comment (PiotrKozlowski) [reply src/cnxml/se.tsx]: https://github.com/openstax-poland/cnx-designer/pull/117#discussion_r571933375

File: src/cnxml/jsx.ts [hunk 8]
```diff
@@ -42,15 +42,15 @@ export interface ProcessingInstruction {
...
 export const Node = {
-    nodeIsNotJSXElement(node: Node):
+    isNotJSXElement(node: Node):
         node is ProcessingInstruction | globalThis.Node | string | Node[] | null {
         return typeof node === 'string'
             || node === null
             || Array.isArray(node)
             || node instanceof globalThis.Node
-            || Node.nodeIsProcessingInstruction(node)
+            || Node.isProcessingInstruction(node)
```
Comment (aiwenar) [inline src/cnxml/jsx.ts]: I would prefer a positive test (`isElement`). It will be simpler to read and won't need to be updated if we ever add another type of `Node`.

File: src/cnxml/se.tsx (hunk 6 above)
Comment (aiwenar) [reply src/cnxml/se.tsx]: I see. In this case please keep those checks in their original order when possible, this will make the diff smaller.

Comment (PiotrKozlowski) [reply src/cnxml/se.tsx]: In this case, I thought that it would be simpler to create `nodeIsNotJSXElement` than creating an if statement with 4 branches and possible with new branches if we add new types to `Node`, but this required changing the order

File: src/cnxml/jsx.ts [hunk 9]
```diff
@@ -29,8 +30,28 @@ export interface Element {
...
+
+export const Node = {
+    isElement(node: Node): node is Element {
+        return Boolean(node
+            && !Array.isArray(node)
+            && !Node.isProcessingInstruction(node)
+            && !(node instanceof globalThis.Node))
+    },
+    isProcessingInstruction(node: Node): node is ProcessingInstruction {
+        return Boolean(node && typeof (node as ProcessingInstruction).target === 'string')
```
Comment (aiwenar) [inline src/cnxml/jsx.ts:52]: Please don't use `Boolean(expression)` to check for truthness. If you want to check that `node` is not `null` do so with `node != null`.

`isElement` should check that `node` is an `Element` directly, rather than checking that it is _not_ any of the other possible types.

Comment (PiotrKozlowski) [reply src/cnxml/jsx.ts:52]: About the check for `isElement` - do you want me to check if the `node` is an object that contains `name` and that the name is an object that contains local of type string and that the node contains `attributes` of type object and also `children` that are of type `node` (which will require creating `Node.isNode` function)?

Comment (aiwenar) [reply src/cnxml/jsx.ts:52]: I don't think we need a full check, after all you are taking `node: Node`. I think something equivalent to `isPlainObject(node) && typeof node.name === 'object'` would suffice (using `isPlainObject` from package `is-plain-object`, it's already in our dependency tree as Slate uses it).

**Review by aiwenar — APPROVED** [review]

**Comment by PiotrKozlowski** [comment]
In order to fix tests I've add `{ text: '' }` to `children` for
```ts
case Node.PROCESSING_INSTRUCTION_NODE:
            editor.apply({
                type: 'insert_node',
                path: path.current!,
                node: {
                    type: 'processing_instruction',
                    target: (child as ProcessingInstruction).target,
                    value: (child as ProcessingInstruction).data,
                    children: [{ text: '' }],
                },
            })
            break
```
